src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/IRazorDocumentMappingServiceExtensions.cs

```diff
@@ -0,0 +1,68 @@
+        => service.TryMapToHostDocumentRange(generatedDocument, projectedRange, MappingBehavior.Strict, out originalRange);
+
+    public static async Task<DocumentPositionInfo?> TryGetPositionInfoAsync(this IRazorDocumentMappingService service, DocumentContext documentContext, Position position, ILogger logger, CancellationToken cancellationToken)
+    {
+        var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        if (!position.TryGetAbsoluteIndex(sourceText, logger, out var absoluteIndex))
+        {
+            return null;
+        }
+
+        return await GetProjectionAsync(service, documentContext, absoluteIndex, cancellationToken).ConfigureAwait(false);
+    }
+
+    public static async Task<DocumentPositionInfo> GetProjectionAsync(this IRazorDocumentMappingService service, DocumentContext documentContext, int hostDocumentIndex, CancellationToken cancellationToken)
+    {
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+
+        sourceText.GetLineAndOffset(hostDocumentIndex, out var line, out var character);
+        var position = new Position(line, character);
+
+        var languageKind = service.GetLanguageKind(codeDocument, hostDocumentIndex, rightAssociative: false);
+        if (languageKind is not RazorLanguageKind.Razor)
+        {
+            var generatedDocument = languageKind is RazorLanguageKind.CSharp
+                ? (IRazorGeneratedDocument)codeDocument.GetCSharpDocument()
+                : codeDocument.GetHtmlDocument();
+            if (service.TryMapToGeneratedDocumentPosition(generatedDocument, hostDocumentIndex, out var mappedPosition, out _))
+            {
+                // For C# locations, we attempt to return the corresponding position
+                // within the projected document
+                position = mappedPosition;
+            }
+            else
+            {
+                // It no longer makes sense to think of this location as C# or Html, since it doesn't
+                // correspond to any position in the projected document. This should not happen
+                // since there should be source mappings for all the C# spans.
+                languageKind = RazorLanguageKind.Razor;
+            }
+        }
+
+        return new DocumentPositionInfo(languageKind, position, hostDocumentIndex);
+    }
+}
+
+/// <summary>
+/// Represents a position in a document. If <see cref="LanguageKind"/> is Razor then the position will be
+/// in the host document, otherwise it will be in the corresponding generated document.
+/// </summary>
+internal record DocumentPositionInfo(RazorLanguageKind LanguageKind, Position Position, int HostDocumentIndex);
```

Comment:
Since this is a core type to the internal API, it probably be defined in its own file.